diff --git a/xf86drm.c b/xf86drm.c
index a8e2bb6..d5f5f02 100644
--- a/xf86drm.c
+++ b/xf86drm.c
@@ -4294,6 +4294,8 @@ static int drmParseOFBusInfo(int maj, int min, char *fullname)
 {
 #ifdef __linux__
     char path[PATH_MAX + 1], *name, *tmp_name;
+    char link[PATH_MAX + 1];
+    ssize_t link_len;

     snprintf(path, sizeof(path), "/sys/dev/char/%d:%d/device", maj, min);

@@ -4301,15 +4303,15 @@ static int drmParseOFBusInfo(int maj, int min, char *fullname)
     tmp_name = name;
     if (!name) {
         /* If the device lacks OF data, pick the MODALIAS info */
-        name = sysfs_uevent_get(path, "MODALIAS");
-        if (!name)
-            return -ENOENT;
-
-        /* .. and strip the MODALIAS=[platform,usb...]: part. */
-        tmp_name = strrchr(name, ':');
-        if (!tmp_name) {
-            free(name);
-            return -ENOENT;
+        link_len = readlink(path, link, sizeof(link) - 1);
+        if (link_len > 0) {
+            link[link_len] = '\0';
+            tmp_name = strrchr(link, '/');
+            if (tmp_name) {
+                tmp_name = strdup(tmp_name + 1);
+            }
         }
-        tmp_name++;
+        if (!tmp_name)
+            return -ENOENT;
+        name = tmp_name;
     }

     strncpy(fullname, tmp_name, DRM_PLATFORM_DEVICE_NAME_LEN);
@@ -4332,6 +4334,8 @@ static int drmParseOFDeviceInfo(int maj, int min, char ***compatible)
     char path[PATH_MAX + 1], *value, *tmp_name;
     unsigned int count, i;
     int err;
+    char link[PATH_MAX + 1];
+    ssize_t link_len;

     snprintf(path, sizeof(path), "/sys/dev/char/%d:%d/device", maj, min);

@@ -4351,19 +4355,16 @@ static int drmParseOFDeviceInfo(int maj, int min, char ***compatible)
         value = sysfs_uevent_get(path, "OF_COMPATIBLE_%u", i);
         tmp_name = value;
         if (!value) {
-            /* If the device lacks OF data, pick the MODALIAS info */
-            value = sysfs_uevent_get(path, "MODALIAS");
-            if (!value) {
-                err = -ENOENT;
-                goto free;
-            }
-
-            /* .. and strip the MODALIAS=[platform,usb...]: part. */
-            tmp_name = strrchr(value, ':');
-            if (!tmp_name) {
-                free(value);
-                return -ENOENT;
+            link_len = readlink(path, link, sizeof(link) - 1);
+            if (link_len > 0) {
+                link[link_len] = '\0';
+                tmp_name = strrchr(link, '/');
+                if (tmp_name) {
+                    tmp_name = strdup(tmp_name + 1);
+                }
             }
-            tmp_name = strdup(tmp_name + 1);
-            free(value);
+            if (!tmp_name) {
+                err = -ENOENT;
+                goto free;
+            }
         }

         (*compatible)[i] = tmp_name;
